### PR TITLE
Coconsensus contract implementation

### LINKED ISSUE #197: Implement Coconsensus contract
1. Implement `Coconsensus` contract 
2. Implement structs, enums and storage variables
    Ref: [Coconsensus PM document](https://github.com/mosaicdao/mosaic-pm/blob/master/specification-discussions/20200108-coconsensus.md)

### PR DESCRIPTION
**Pull request by deepesh-kn**
This pull request contains the initial implementation of `Coconsenus` contract
fixes: #197

### CODE REVIEW

File: contracts/coconsensus/GenesisCoconsensus.sol [hunk 1]
```diff
@@ -32,17 +32,23 @@ contract GenesisCoconsensus {
...
+    /** Link list of metachain ids (for all protocores). */
     mapping(bytes32 => bytes32) public genesisMetachainIds;
 
-    /** Mapping of metablock id to its core contract address on origin chain. */
-    mapping(bytes32 => address) public genesisCores;
+    /** Mapping of metachain id to its anchor contract address on auxiliary chain. */
+    mapping(bytes32 => address) public genesisAnchors;
 
-    /** Mapping of metablock id to the epoch lengths. */
+    /** Mapping of metachain id to its protocore contract address on auxiliary chain. */
```
Comment (abhayks1) [inline contracts/coconsensus/GenesisCoconsensus.sol]: documentation is going more than 79 char limit.

File: contracts/coconsensus/Coconsensus.sol [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    /* Storage */
+
+    /** Mapping to track the blocks for each metachain. */
+    mapping (bytes32 /* metachainid */ => mapping(uint256 /* blocknumber */ => Block)) blockchains;
+
+    /** Mapping of metachain id to latest block number(tip) stored in blockchains. */
+    mapping (bytes32 /* metachainId */ => uint256 /* blocknumber */) blockTips;
+
+    /** Mapping of metachain id to the protocore contract address. */
+    mapping (bytes32 /* metachainId */ => Protocore) protocores;
```
Comment (abhayks1) [inline contracts/coconsensus/Coconsensus.sol]:
```suggestion
    mapping (bytes32 /* metachainId */ => ProtocoreI) protocores;
```
We can use Protocore interface here. We have followed similar pattern in Consensus contract.

File: contracts/coconsensus/Coconsensus.sol [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    /** Mapping of metachain id to the domain separators. */
+    mapping (bytes32 /* metachainId */ => bytes32 /* domain separator */) domainSeparators;
+}
```
Comment (abhayks1) [inline contracts/coconsensus/Coconsensus.sol:68]: Looking at the coconsensus specification doc, `anchors` mapping also needs to be there.
https://github.com/mosaicdao/mosaic-pm/blob/master/specification-discussions/20200108-coconsensus.md

**Review by abhayks1 — CHANGES_REQUESTED** [review]
There are few inline comments.

File: contracts/coconsensus/GenesisCoconsensus.sol (hunk 1 above)
Comment (deepesh-kn) [reply contracts/coconsensus/GenesisCoconsensus.sol]: Made it multiline.

File: contracts/coconsensus/Coconsensus.sol (hunk 2 above)
Comment (deepesh-kn) [reply contracts/coconsensus/Coconsensus.sol]: Done.

File: contracts/coconsensus/Coconsensus.sol (hunk 3 above)
Comment (deepesh-kn) [reply contracts/coconsensus/Coconsensus.sol:68]: Added observers mapping.

File: contracts/coconsensus/Coconsensus.sol [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        uint256 statusDynasty;
+    }
+
+
+    /* Storage */
+
+    /** Mapping to track the blocks for each metachain. */
+    mapping (bytes32 /* metachainId */ => mapping(uint256 /* blocknumber */ => Block)) blockchains;
+
+    /** Mapping of metachain id to latest block number(tip) stored in blockchains. */
```
Comment (abhayks1) [inline contracts/coconsensus/Coconsensus.sol]: title and this line doc is more than 79 chars.

**Review by abhayks1 — APPROVED** [review]
LGTM 👍 
Minor change requested

File: contracts/coconsensus/Coconsensus.sol (hunk 4 above)
Comment (deepesh-kn) [reply contracts/coconsensus/Coconsensus.sol]: Done

**Pull request merged**